### PR TITLE
fix(rocjitsu): honor guest MODE.FP_ROUND in host arithmetic

### LINKED ISSUE #10855: [rocjitsu] MODE.FP_ROUND does not reach the arithmetic: every directed rounding mode returns the round-to-nearest result
## Summary

`rocjitsu` executes the guest's `MODE.FP_ROUND` by setting the **host's** rounding mode — `fp_mode.h` wraps each architectural operation in a `ScopedFenv` (`feholdexcept` + `fesetround`) and lets the hardware round.

Nothing tells the compiler that rounding is dynamic, so it assumes the default mode and is free to move the arithmetic across the `fesetround`. **Every directed rounding mode silently produces the round-to-nearest answer.**

## Reproducer

`V_FMAC_F64` computing `1.0 * 2^-53 + 1.0` — an exact tie, half an ulp above 1.0, so the rounding mode alone decides the result.

| `MODE.FP_ROUND` | guest should see | scalar path returns |
|---|---|---|
| 0 nearest-even | `0x3ff0000000000000` | `0x3ff0000000000000` ✅ |
| 1 +inf | `0x3ff0000000000001` | `0x3ff0000000000000` ❌ |
| 2 −inf / 3 zero | `0x3ff0000000000000` | wrong on operands past the halfway point ❌ |

Independent of the test suite — a program calling `fp_mode::fma_f64` with inputs taken from `argv` so nothing can be constant-folded:

```
$ g++-13 -O3 …                  round=1 -> 0x3ff0000000000000   # wrong
$ g++-13 -O3 -frounding-math …  round=1 -> 0x3ff0000000000001   # correct
```

## Not compiler-specific

Measured on the same probe, wrong without `-frounding-math` and correct with it, on **all** of:

- GCC 13.4
- clang 14
- clang 15
- **AMD clang 22** (`/opt/rocm/llvm/bin/clang++`)

So building with amdclang does not avoid it.

## Impact

Any guest code that sets a directed rounding mode gets nearest-even instead. The scalar and SIMD paths are affected differently, so a result can also depend on `RJ_FORCE_SCALAR`.

It currently shows up as three failing tests on `develop`:

```
Vop2FmaF64SimdCorrectness.AllRoundAndDenormModesMatchScalar
Vop2FmaF64SimdCorrectness.SpecialValuesAndDenormBoundariesMatchScalarExactly
Gfx1250ExecutionTest.FusedOperationsHonorF16F64ModeControls
```

Those assert `simd == scalar`, which catches the divergence but names whichever path it is handed as the reference — here the **scalar** path is the wrong one, so the failure reads as "the two disagree" rather than "one is wrong".

## Note

This is distinct from #10330, which is about scalar and SIMD matrix kernels accumulating MACs differently. Same visible symptom (a result depending on `RJ_FORCE_SCALAR`), different cause.

### PR DESCRIPTION
**Pull request by lialan**
Closes #10855

## The bug

`MODE.FP_ROUND` is a guest architectural register: an AMDGPU wave can select
round-to-nearest-even, +inf, -inf, or toward-zero, and its FP results must honor
that choice.

RocJITsu implements the operations that need more precision than their guest
type by borrowing the host FPU. `fp_mode.h` wraps this arithmetic in a
`ScopedFenv`, which sets the requested host rounding mode and restores the
caller's environment afterward. Without a strict floating-point compiler
option, however, the compiler may assume the host rounding mode never changes
and move or fold the arithmetic as round-to-nearest.

For F64, `v_fmac_f64` of `1.0 * 2^-53 + 1.0` demonstrates the wrong answer:

| | round-to-nearest | +inf | -inf | toward-zero |
|---|---|---|---|---|
| hardware / ISA | `1.0` | `1.0 + 1ulp` | `1.0` | `1.0` |
| RocJITsu, before | `1.0` | **`1.0`** | `1.0` | `1.0` |
| RocJITsu, after | `1.0` | `1.0 + 1ulp` | `1.0` | `1.0` |

Review also identified the same class of bug in F16 FMA. F16 operands are
represented exactly as doubles, but their exact product-plus-addend can require
more than the 53 bits available in a double. For example,
`0x0001 * 0x0001 + 0x7800` under guest round-toward-+inf must produce `0x7801`.
If the double FMA first runs under host round-to-nearest, the tiny product is
lost before the software F16 rounding step and the result is incorrectly
`0x7800`. Both the scalar and SIMD helpers had this issue.

## The fix

- Apply the strict rounding option only to targets that can inline the
  host-environment helpers: `-frounding-math` for GNU-style GCC/Clang and
  `/fp:strict` for MSVC, including clang-cl. The MSVC check comes first because
  clang-cl satisfies both `MSVC` and the Clang compiler-ID test.
- Split helpers that never touch the host environment into
  `fp_mode_no_fenv.h`, so unrelated translation units do not inherit the strict
  option.
- Compute scalar and SIMD F16 FMA double intermediates inside `ScopedFenv`, then
  restore the caller's environment before applying the software F16 result
  policy.
- Guard the fenv-sensitive header with the corresponding build definition so a
  new include site cannot silently compile without strict rounding semantics.

## Tests

- F64 scalar and SIMD directed-rounding cases are pinned to ISA-derived bit
  patterns rather than only compared with each other.
- `FpModePolicyTest.F16FmaUsesGuestRoundingAndRestoresAmbientHostEnvironment`
  exercises scalar and SIMD F16 counterexamples for +inf and -inf, a mismatched
  ambient/guest round-to-nearest case, and restoration of the caller's host
  rounding mode.
- Release GCC 13 default build completes successfully.
- Focused CTest: 14/14 passed.
- Full amdisa Python suite: 2,074 passed, 46 skipped.
- PR-range pre-commit hooks and Ruff pass.

The clang-cl branch is selected correctly by CMake; runtime validation here is
on Linux/GCC because this host does not provide a Windows clang-cl environment.

### CODE REVIEW

**Review by atgutier — COMMENTED** [review]
- MAJOR — Windows/clang-cl retains the original rounding bug
    emulation/rocjitsu/CMakeLists.txt:63

    The new condition excludes MSVC, but provides no equivalent strict-rounding option. Consequently,
    MSVC and clang-cl builds still compile ScopedFenv callers without a dynamic-rounding contract.

    I reproduced this with AMD clang-cl 23 targeting x86_64-pc-windows-msvc. At -O2, the compiler
    moved fma after fesetenv, so it executes under the restored rounding mode—the exact bug this PR
    fixes on Linux. With /fp:strict, the fma remained between fesetround and fesetenv; clang-cl
    translates that option to -frounding-math -ffp-exception-behavior=strict.

    Fix direction:

    if(MSVC)
        add_compile_options(/fp:strict)
    elseif(CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
        add_compile_options(-frounding-math)
    endif()

    Add a Windows/compiler regression if that configuration is supported. The PR’s Windows CI job was
    skipped.

  - MINOR — the strict-rounding option is applied to the entire project and its dependencies
    emulation/rocjitsu/CMakeLists.txt:70

    add_compile_options(-frounding-math) is set before FetchContent_MakeAvailable and every project
    add_subdirectory. GoogleTest, FlatBuffers, simdojo, VM, DBT, tools, tests, and unrelated ISA code
    therefore inherit an optimization-constraining floating-point contract, even though host-rounding
    changes occur at four AMDGPU helper sites.

    This matters because the PR already measures approximately 6% lower emulation throughput. Header-
    only code does require every calling translation unit to use compatible semantics, but it does not
    require every target and dependency to do so.

    Prefer moving the fenv-sensitive helpers out of line into strict-rounding translation units. As an
    intermediate step, apply the option only to the AMDGPU execution targets and direct helper tests.
    Confidence is high that the current scope is excessive; how much performance narrowing recovers
    needs measurement.

  - NIT — the new test can report scalar fallback as SIMD coverage
    emulation/rocjitsu/tests/simd_correctness/vop2_fma_f64_simd_correctness_test.cpp:233

    The test only checks util::has_stdx_simd, then labels run_fmac(false) as SIMD. On clang + libstdc+
    + + AVX-512 configurations, UTIL_SIMD_BROKEN_NATIVE_64BIT_MASKS disables the VOP2 f64 SIMD probe
    completely. Execution falls through to scalar, so both “scalar” and “simd” assertions can test the
    scalar implementation.

    Gate the SIMD assertion on that workaround as well, or invoke the SIMD probe directly and assert
    that it actually handled the instruction.

**Comment by lialan** [comment]
@atgutier Both fixed: `-frounding-math` is now carried in `RJ_STRICT_FP_ROUNDING_OPTIONS` and applied only to the AMDGPU `rocjitsu_isa_<arch>_exec` object libraries and to `tests/` (805/805 TUs down to 396/805 — googletest, flatbuffers, simdojo, the VM, the DBT, tools and the RISC-V ISA are all out; emptying the variable makes the rounding tests fail again, so that scope is still the load-bearing one), and the f64 test now gates its SIMD half on a new `util::has_stdx_simd_64bit_lanes` so the scalar fallback is no longer counted as SIMD coverage. Build 818/818 and ctest 3689/3689 on g++-13 Release; I don't expect this to recover the 6% since the `_exec` TUs keep the same flags, and I dropped the MSVC guard because we have never built with MSVC or clang-cl and that CI job was skipped rather than run.

File: emulation/rocjitsu/CMakeLists.txt [hunk 1]
```diff
@@ -45,6 +45,28 @@ set(CMAKE_C_VISIBILITY_PRESET hidden)
...
+# it, so the flag is what is portable between the two. It only forbids
+# optimisations that assume a rounding mode, so it cannot make arithmetic less
+# correct. It has to be global because fp_mode.h is header-only and inlines into
+# every ISA translation unit.
+if(
+    CMAKE_CXX_COMPILER_ID STREQUAL "GNU"
+    OR CMAKE_CXX_COMPILER_ID MATCHES "Clang"
+)
+    add_compile_options(-frounding-math)
+endif()
```
[the remark this answers is not in the document]
Comment (lialan) [reply emulation/rocjitsu/CMakeLists.txt]: Reverted this rather than keeping it. rocjitsu is not built with MSVC or clang-cl here, and the Windows CI job on this PR was skipped rather than run, so the guard was hardening a configuration that has never been tried. The flag is now applied via target_compile_options on the AMDGPU _exec object libraries and at tests/ scope instead of globally, so the compiler-ID check is no longer the only thing bounding it. If Windows becomes a supported configuration this should come back as a real if(MSVC) branch with a regression test, not as an unexercised guard.

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/isa/CMakeLists.txt [hunk 2]
```diff
@@ -50,6 +50,12 @@ function(rj_add_split_amdgpu_isa_library arch)
     rj_add_object_library(${_target}_model ${_model_sources})
     rj_add_object_library(${_target}_exec ${_execution_sources})
 
+    # The execution sources are the ones that include fp_mode.h and simd_glue.h,
+    # so they are the ones that need the compiler to stop assuming a rounding
+    # mode across the ScopedFenv. See RJ_STRICT_FP_ROUNDING_OPTIONS in the
+    # top-level CMakeLists.txt. The _model sources never see either header.
+    target_compile_options(${_target}_exec PRIVATE ${RJ_STRICT_FP_ROUNDING_OPTIONS})
```
Comment (kuhar) [inline emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/isa/CMakeLists.txt]: Please run gersemi on this invocation. gersemi 0.25.1 --check reports that emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/isa/CMakeLists.txt:57 would be reformatted, which matches the current pre-commit failure.

File: emulation/rocjitsu/lib/util/include/util/simd.h [hunk 3]
```diff
@@ -47,6 +47,15 @@ inline constexpr bool has_stdx_simd =
     false;
 #endif
 
+/// Compile-time switch for the 64-bit-lane SIMD paths specifically. The
+/// `ROCJITSU_TRY_SIMD_*_F64` probes in simd_glue.h expand to nothing when
+/// `UTIL_SIMD_BROKEN_NATIVE_64BIT_MASKS` is set, so execution falls through to
+/// the scalar implementation even though `has_stdx_simd` is true. A test that
+/// gates only on `has_stdx_simd` would then run the scalar path twice and
+/// report the second run as SIMD coverage; gate on this instead.
+inline constexpr bool has_stdx_simd_64bit_lanes =
```
Comment (kuhar) [inline emulation/rocjitsu/lib/util/include/util/simd.h:57]: It would be good to use `has_stdx_simd_64bit_lanes` consistently for 64-bit SIMD coverage. The four older cases in `vop2_fma_f64_simd_correctness_test.cpp` still gate only on `has_stdx_simd`, so when the native-mask workaround removes `ROCJITSU_TRY_SIMD_VOP2_FMA_F64`, each runs the scalar implementation twice. `tests/cdna5_execution_test.cpp:1050` also open-codes the same capability check; using the helper there would keep this util-level contract centralized. The helper comment should describe that capability generically rather than naming an AMDGPU probe.

File: emulation/rocjitsu/tests/simd_correctness/vop2_fma_f64_simd_correctness_test.cpp [hunk 4]
```diff
@@ -211,4 +211,79 @@ TEST(Vop2FmaF64SimdCorrectness, SpecialValuesAndDenormBoundariesMatchScalarExact
...
+  struct Case {
+    const char *what;
+    uint64_t src0;
+    uint64_t src1;
+    uint64_t accumulator;
+    // Indexed by MODE.FP_ROUND: 0 nearest-even, 1 +inf, 2 -inf, 3 zero.
+    std::array<uint64_t, 4> expected;
+  };
+
+  const Case cases[] = {
```
Comment (kuhar) [inline emulation/rocjitsu/tests/simd_correctness/vop2_fma_f64_simd_correctness_test.cpp:271]: We should add a negative inexact case so this test distinguishes round-toward-negative-infinity from round-toward-zero. Both current rows are positive, so MODE values 2 and 3 have identical expected bits and a swap of `FE_DOWNWARD` and `FE_TOWARDZERO` would still pass. A negative mirror of the 0.5625-ulp row would require the more-negative neighbor for nearest/downward and the truncated value for upward/zero, independently exercising all four mappings.

File: emulation/rocjitsu/CMakeLists.txt [hunk 5]
```diff
@@ -45,6 +45,48 @@ set(CMAKE_C_VISIBILITY_PRESET hidden)
...
+# correct.
+#
+# It is carried as a variable rather than applied here, because the contract it
+# imposes is only needed where fp_mode.h's ScopedFenv actually inlines. Applying
+# it at this scope would reach googletest, flatbuffers, simdojo, util, the VM,
+# the DBT and every tool as well, constraining code that never touches the host
+# rounding mode. Two places consume it, and they are the closure of the header:
+#
+#   - rocjitsu_isa_<arch>_exec (isa/CMakeLists.txt), the object libraries that
+#     own the generated *_exec*.cpp sources. Those are the only translation
```
Comment (kuhar) [inline emulation/rocjitsu/CMakeLists.txt:70]: This would be safer if the header boundary matched the compiler-flag boundary. `compute_unit.cpp` includes `fp_mode.h` through `alu_exceptions.h`, but its compile command lacks `-frounding-math` while the `_exec` translation units have it. It currently uses only `effective_omod`, so no fenv-sensitive helper is emitted there, but a future call to `fma_f64` or `finish_f64` could create an unflagged inline definition in the same image. Please split non-fenv helpers such as `effective_omod` out of `fp_mode.h`, or otherwise make it structural that fenv-sensitive helpers cannot be compiled without the strict-rounding option.

**Review by kuhar — COMMENTED** [review]
Added some new suggestions

File: emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/isa/CMakeLists.txt (hunk 2 above)
Comment (lialan) [reply emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/isa/CMakeLists.txt]: Done in 96285d2 — ran gersemi 0.25.1 and it wraps the call. `gersemi --check` is clean on all three CMakeLists in the change now.

File: emulation/rocjitsu/lib/util/include/util/simd.h (hunk 3 above)
Comment (lialan) [reply emulation/rocjitsu/lib/util/include/util/simd.h:57]: Done in 0bb46ef.

- The four older cases in `vop2_fma_f64_simd_correctness_test.cpp` now gate on the helper. Two of them (`AllRoundAndDenormModesMatchScalar`, `SpecialValuesAndDenormBoundariesMatchScalarExactly`) are purely relative, so they skip. The other two also pin an absolute value — the decoded literal high word, and inactive lanes surviving a partial EXEC — so I kept those assertions running unconditionally against the scalar path and gated only the comparison. The old whole-test skip was discarding that coverage on the affected configurations.
- `cdna5_execution_test.cpp:1050` uses the helper.
- Comment restated in terms of the capability rather than the AMDGPU probes.

File: emulation/rocjitsu/tests/simd_correctness/vop2_fma_f64_simd_correctness_test.cpp (hunk 4 above)
Comment (lialan) [reply emulation/rocjitsu/tests/simd_correctness/vop2_fma_f64_simd_correctness_test.cpp:271]: Done in d30f3e2. Added the negative mirror of the 0.5625-ulp row: nearest and -inf take the more-negative neighbor, +inf and zero truncate toward -1.25, so both pairs split and each MODE value maps to exactly one host mode.

Checked that it actually discriminates rather than assuming it: with `FE_DOWNWARD` and `FE_TOWARDZERO` transposed in `host_round_mode`, the suite fails on that row alone, at round=2 and round=3, on both the scalar and the SIMD path. The two positive rows still pass under the transposition, which is the gap you identified.

Expected bits come from the ISA definition and were cross-checked against a standalone `fesetround`/`fma` probe built with `-frounding-math`, not against rocjitsu.

File: emulation/rocjitsu/CMakeLists.txt (hunk 5 above)
Comment (lialan) [reply emulation/rocjitsu/CMakeLists.txt:70]: Done in fdf9f0a — took the split rather than the "otherwise", since the guard only means something once the header contains nothing but fenv-sensitive code.

`fp_mode_no_fenv.h` now holds `effective_omod`, `effective_f16_omod`, the `finalize_omod_*` family and the input modifiers: integer and bit-pattern work only, correct under any FP settings, no requirement attached. `alu_exceptions.h` includes that, so `compute_unit.cpp` can no longer see `fma_f64` or `finish_f64` at all — the future call you describe is now a compile error at the include, not an unflagged inline definition in the image.

What is left in `fp_mode.h` demands the contract: `RJ_STRICT_FP_ROUNDING_OPTIONS` carries `-DROCJITSU_STRICT_FP_ROUNDING` next to the flag, and the header `#error`s without it, pointing at the other header. So the two boundaries cannot drift apart — a target that gets the header without the option fails to build. Verified both directions by hand (bare `#include` fails with the message; with the define and flag it compiles).

`fma_f16` stays on the strict side. It rounds to F16 in software, but the intermediate is a double `std::fma` that is not exact at that width for every input pair, so the ambient rounding mode can still reach it.

Also filled in the branches the variable was missing: `/fp:strict` for MSVC, and a `message(WARNING)` instead of silence for a compiler with no known spelling.

ctest 3689/3689 on this branch.

File: emulation/rocjitsu/lib/python/amdisa/tests/test_generator_profile_gates.py [hunk 6]
```diff
@@ -38,6 +38,12 @@
 from amdisa.codegen.execute.simd_codegen import simd_probe_line
 from amdisa.cross_isa import CrossIsaAnalyzer
 from amdisa.gpuisa import Instruction, Operand
+
+# The generated *_exec.cpp sources reach fp_mode.h, which refuses to compile
+# without the strict rounding-mode build options. CMake attaches these to
+# rocjitsu_isa_<arch>_exec via RJ_STRICT_FP_ROUNDING_OPTIONS; a preprocess-only
+# check has to supply the define the same way. See emulation/rocjitsu/CMakeLists.txt.
+STRICT_FP_ROUNDING_DEFINE = '-DROCJITSU_STRICT_FP_ROUNDING=1'
```
Comment (kuhar) [inline emulation/rocjitsu/lib/python/amdisa/tests/test_generator_profile_gates.py]: Please move STRICT_FP_ROUNDING_DEFINE below the complete import block. ruff check lib/python/amdisa/tests/test_generator_profile_gates.py reports I001 at line 47 because the constant currently splits the amdisa imports.

**Review by kuhar — COMMENTED** [review]
We should apply guest rounding to the double intermediate in both F16 FMA paths, or revise the new header contract and keep these helpers outside the fenv-sensitive boundary. At emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/isa/arch/amdgpu/shared/fp_mode.h:98, fma_f16 passes round_mode only to the software F16 conversion, while std::fma runs under the ambient host mode; emulation/rocjitsu/lib/rocjitsu/src/rocjitsu/isa/arch/amdgpu/shared/simd_glue.h:548 has the same issue. With src0=src1=0x0001, src2=0x7800, and guest round-up, a focused probe returned 0x7800 under ambient round-to-nearest and 0x7801 under ambient round-up. The architectural result therefore still depends on unrelated host state despite the file-level MODE.FP_ROUND contract.

File: emulation/rocjitsu/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (kuhar) [reply emulation/rocjitsu/CMakeLists.txt]: We should make the MSVC-style branch take precedence over the compiler-ID check. On the current head, clang-cl sets both MSVC and CMAKE_CXX_COMPILER_ID=Clang, so it still reaches -frounding-math instead of /fp:strict; this is the same configuration discussed in this thread. Testing MSVC first, or excluding MSVC from the Clang arm, keeps the new branch effective.

**Review by kuhar — CHANGES_REQUESTED** [review]
I found [more simulator bugs](https://gist.github.com/kuhar/5a94a0c030f6aaeda3e69e635f64f13a) in this area while following up on recent decoder fixes, prioritizing gfx1250 and gfx950. Ordinary VALU and VOPD/VOPD3 instructions were missing guest rounding and denormal handling, and we also found separate SDWA scaling and clamp defects. We need to expand the work you are doing here to cover those additional execution paths.

The fixes are complementary: our ordinary-arithmetic branch still fails two of the F16 FMA cases added in this PR, while this PR doesn't address the missing MODE handling in those ordinary arithmetic paths. I'd like us to combine the coverage around an **explicit-policy arithmetic API**:

- Pass operand bits and guest FP policy explicitly.
- Return result bits and exception status explicitly.
- Implement rounding without changing the host FP environment or requiring strict-FP compiler flags.
- Preserve sufficient precision through output modifiers, rounding at the architecturally required stage.
- Add optimized paths where we can establish independence from ambient host state.

`ScopedFenv` restores the caller's environment, but execution still temporarily changes ambient thread state and depends on the compiler respecting those transitions. Our current patch expands that dependency too, so both implementations need reconsideration here.

I tried an external APFloat prototype: 256 cases pass on both GCC and Clang without strict-FP flags, while preserving host rounding, exception flags, FTZ, and DAZ. That demonstrates a possible correctness path, although performance and complete AMD-specific behavior still need evaluation.

I suggest keeping both sets of regressions and using them to qualify that shared backend as we expand the instruction coverage.

File: emulation/rocjitsu/lib/python/amdisa/tests/test_generator_profile_gates.py (hunk 6 above)
Comment (lialan) [reply emulation/rocjitsu/lib/python/amdisa/tests/test_generator_profile_gates.py]: Fixed in `f7eb6b75ac`: `STRICT_FP_ROUNDING_DEFINE` now follows the complete `amdisa` import block. Ruff and PR-range pre-commit pass, and the full amdisa suite reports 2,074 passed and 46 skipped.

File: emulation/rocjitsu/CMakeLists.txt (hunk 1 above)
[the remark this answers is not in the document]
Comment (lialan) [reply emulation/rocjitsu/CMakeLists.txt]: Fixed in `f7eb6b75ac`: the `MSVC` arm now comes first, so clang-cl selects `/fp:strict`, while GCC and GNU-style Clang retain `-frounding-math`. The Linux/GCC Release build passed; no Windows clang-cl runtime was available here, so validation of that arm is structural rather than execution-based.

**Comment by lialan** [comment]
@kuhar Addressed the remaining review points in `f7eb6b75ac`.

Both scalar `fma_f16` and SIMD `fma_f16_mode_simd` now compute the double FMA inside `ScopedFenv(round_mode)` and restore the caller's environment before software F16 handling. The new `FpModePolicyTest.F16FmaUsesGuestRoundingAndRestoresAmbientHostEnvironment` covers your `0x0001 * 0x0001 + 0x7800` guest-round-up reproducer, an inverse ambient/guest mismatch, a negative round-down case, scalar and SIMD paths, and host-mode restoration.

I also moved `STRICT_FP_ROUNDING_DEFINE` below the complete import block and made the `MSVC` arm take precedence so clang-cl selects `/fp:strict`, while GCC and GNU-style Clang retain `-frounding-math`.

Focused CTest passed 14/14; the Release GCC 13 build, full amdisa suite (2,074 passed, 46 skipped), Ruff, and PR-range pre-commit passed. GitHub CI is terminal with no failures. This host lacks a Windows clang-cl environment, so validation of that arm is structural rather than execution-based.

Could you please re-review?

**Comment by lialan** [comment]
Correction: your new changes-requested review landed while the last CI job was running and was not present in my earlier thread audit. I see the broader explicit-policy arithmetic request now. Please hold re-review; I am auditing the linked ordinary VALU/VOPD regressions and the host-`fenv` dependency before asking again.
